feat(ivy): support generation of flags for directive injection

mhevery opened the pull request:
This change changes:
- compiler uses `directiveInject` instead of `inject` for `Directive`s
- unifies the flags in `di` as well as `render3`
- changes the signature of `directiveInject` to match `inject` In prep for #23330
- compiler now generates flags for injection.

Compiler portion of #23342
Prep for #23330 

feat(ivy): support injection even if no injector present

- Remove default injection value from `inject` / `directiveInject` since
  it is not possible to set using annotations.
- Module `Injector` is stored on `LView` instead of `LInjector` data 
  structure because it can change only at `LView` level. (More efficient)
- Add `ngInjectableDef` to `IterableDiffers` so that existing tests can 
  pass as well as enable `IterableDiffers` to be injectable without 
  `Injector`

mhevery committed:
feat(ivy): support generation of flags for directive injection

This change changes:
- compiler uses `directiveInject` instead of `inject` for `Directive`s
- unifies the flags in `di` as well as `render3`
- changes the signature of `directiveInject` to match `inject` In prep for #23330
- compiler now generates flags for injection.

Compiler portion of #23342
Prep for #23330

kara commented on packages/compiler/test/render3/r3_view_compiler_di_spec.ts:
```diff
@@ -0,0 +1,71 @@
+import {MockDirectory, setup} from '../aot/test_util';
+import {compile, expectEmit} from './mock_compile';
+
+fdescribe('compiler compliance: dependency injection', () => {
```
Remove fdescribe?

mhevery replied:
Fixed

kara reviewed (commented):
Looks good to me, except for the fdescribe

kara commented on packages/core/src/render3/di.ts:
```diff
@@ -133,29 +133,6 @@ export function getOrCreateNodeInjectorForNode(node: LElementNode | LContainerNo
- * @param token The token associated with the error
- * @returns The error that was created
- */
-function createInjectionError(text: string, token: any) {
```
Did you mean to remove this? Looks like it's still used in this file.

mhevery replied:
Working on two things in one PR, reverted

kara reviewed (approved):
LGTM

mary-poppins commented:
You can preview 4d6700a at https://pr23345-4d6700a.ngbuilds.io/.

vicb commented on packages/compiler/src/render3/r3_view_compiler.ts:
```diff
@@ -972,10 +969,33 @@ export function createFactory(
             token.identifier != null ? outputCtx.importExpr(tokenRef) : o.literal(tokenRef);
-        args.push(o.importExpr(R3.inject).callFn([value]));
+        const directiveInjectArgs = [tokenValue];
+        let flags = InjectFlags.Default;
```
`const flags = moveThatLogicFurtherDown(dependency);` ?

mhevery replied:
extracted

vicb commented on packages/compiler/test/render3/r3_view_compiler_di_spec.ts:
```diff
@@ -0,0 +1,71 @@
+    };
+
+    // The template should look like this (where IDENT is a wild card for an identifier):
+    const template = `
```
rename `template` to `factory` or `injectxxxyyy` ? (in comment, it and expect)
Use `$r3$` instead of `IDENT` for consistency with other tests ?

mhevery replied:
done

mary-poppins commented:
You can preview 7b8bcef at https://pr23345-7b8bcef.ngbuilds.io/.

mary-poppins commented:
You can preview c75e17f at https://pr23345-c75e17f.ngbuilds.io/.

mary-poppins commented:
You can preview 49cb591 at https://pr23345-49cb591.ngbuilds.io/.

mhevery committed:
feat(ivy): support injection even if no injector present

- Remove default injection value from `inject` / `directiveInject` since
  it is not possible to set using annotations.
- Module `Injector` is stored on `LView` instead of `LInjector` data
  structure because it can change only at `LView` level. (More efficient)
- Add `ngInjectableDef` to `IterableDiffers` so that existing tests can
  pass as well as enable `IterableDiffers` to be injectable without
  `Injector`

mhevery committed:
docs(ivy): Clean up incorrect comments

mhevery committed:
fix(ivy): Update Todo app to take advantage of optional injector

vicb commented on packages/core/test/bundling/todo/index.ts:
```diff
@@ -111,9 +111,7 @@ export class ToDoAppComponent {
-  // TODO(misko) Fix injection
-  // constructor(todoStore: TodoStore) { this.todoStore = todoStore; }
-  constructor() { this.todoStore = new TodoStore(); }
+  constructor(todoStore: TodoStore) { this.todoStore = todoStore; }
```
`constructor(public todoStore: TodoStore) {}` should work now

mhevery replied:
done

mary-poppins commented:
You can preview 493ab9b at https://pr23345-493ab9b.ngbuilds.io/.

kara commented on packages/core/test/render3/di_spec.ts:
```diff
@@ -397,6 +397,35 @@ describe('di', () => {
+      }
+
+      class MyComponent {
+        constructor(public mySerivce: MyService) {}
```
Rename to `myService` ?

mhevery replied:
fixed

kara commented on packages/core/test/render3/di_spec.ts:
```diff
@@ -397,6 +397,35 @@ describe('di', () => {
       expect(log).toEqual(['DirB', 'DirB', 'DirA (dep: DirB - 2)']);
     });
 
+    it('should crete instance even when no injector present', () => {
```
`crete` -> `create`

mhevery replied:
fixed

kara reviewed (approved):
A few small typos, otherwise looks good to me

kara reviewed (approved):
LGTM

mary-poppins commented:
You can preview 091c9e1 at https://pr23345-091c9e1.ngbuilds.io/.

mary-poppins commented:
You can preview e2a3a2f at https://pr23345-e2a3a2f.ngbuilds.io/.

vicb commented on packages/core/src/di/defs.ts:
```diff
@@ -25,8 +25,22 @@ import {ClassProvider, ClassSansProvider, ConstructorProvider, ConstructorSansPr
+   * scope (e.g. `'root'`). A value of `null` indicates that the injectable does not belong to any
+   * scope.
+   */
   providedIn: InjectorType<any>|'root'|'any'|null;
```
add docs for 'any'

mhevery replied:
@alxhub What is `any` for?

mhevery replied:
I read the source code, added appropriate docs.

trotyl replied:
Please note [`defineInjectable`](https://angular.io/api/core/defineInjectable) is also a public API in non-ivy code, but `'any'` [doesn't make any effect here](https://stackblitz.com/edit/angular-vhhidk?file=src%2Fapp%2Fapp.component.ts). Wouldn't that be incorrect for public doc?

vicb commented on packages/core/src/di/defs.ts:
```diff
@@ -107,6 +121,7 @@ export function defineInjectable<T>(opts: {
   return {
     providedIn: (opts.providedIn as InjectorType<any>| 'root' | null | undefined) || null,
```
no 'any' ?

mhevery replied:
added

vicb commented on packages/core/src/di/injector.ts:
```diff
@@ -445,19 +456,21 @@ export function setCurrentInjector(injector: Injector | null): Injector|null {
+  } else if (_currentInjector === null) {
+    const injectableDef: InjectableDef<T> = (token as any).ngInjectableDef;
+    if (injectableDef && injectableDef.providedIn == 'root') {
+      return injectableDef.value === undefined ? injectableDef.value = injectableDef.factory() :
```
do we want to be able to provide `undefined` by using a special value (ie `{}`) instead ?

mhevery replied:
I was thinking about it, and decided that it is not worth the complexity. We can always added later.

vicb commented on packages/core/src/render3/interfaces/view.ts:
```diff
@@ -189,6 +190,11 @@ export interface LView {
   queries: LQueries|null;
+
+  /**
+   * Injector to be used (if any).
```
useful comment ?

mhevery replied:
updated comment.

trotyl commented on packages/core/src/change_detection/differs/iterable_differs.ts:
```diff
@@ -135,6 +137,11 @@ export interface IterableDifferFactory {
  *
  */
 export class IterableDiffers {
+  static ngInjectableDef: InjectableDef<IterableDiffers> = defineInjectable({
```
Why would compilation result lives in source code? Shouldn't it using `@Injectable()` decorator?

mhevery replied:
Few reasons:
1.) we currently don't compile core with ivy so the decorator does not get lowered.
2.) There is a circular dependency between the compiler and core which makes it hard for the compiler to compile tho core. Rather than solving that, we have opted for hand coding it.

alfaproject replied:
This comment should probably be in the source code

mary-poppins commented:
You can preview 685546e at https://pr23345-685546e.ngbuilds.io/.

IgorMinar commented:
this PR changes the API surface. @mhevery let's discuss tomorrow before merging this.

mhevery commented:
Discussed with @IgorMinar and we agreed to merge.

mary-poppins commented:
You can preview 7cf605d at https://pr23345-7cf605d.ngbuilds.io/.

alfaproject commented on tools/public_api_guard/core/core.d.ts:
```diff
@@ -376,8 +376,10 @@ export interface InjectDecorator {
 
 export declare const enum InjectFlags {
     Default = 0,
-    SkipSelf = 1,
```
Why change the order? It's a bitmask anyway o:
